crate/server/src/core/operations/validate.rs

```diff
@@ -0,0 +1,586 @@
+//             self.current = *son_cert_id;
+//             Result::Ok(Some(res.clone()))
+//         } else {
+//             Result::Ok(None)
+//         }
+//     }
+// }
+
+/// This operation requests the server to validate a certificate chain and return
+/// information on its validity.
+/// Only a single certificate chain SHALL be included in each request.
+/// The request MAY contain a list of certificate objects, and/or a list of
+/// Unique Identifiers that identify Managed Certificate objects.
+/// Together, the two lists compose a certificate chain to be validated.
+/// The request MAY also contain a date for which all certificates in the
+/// certificate chain are REQUIRED to be valid.
+/// The method or policy by which validation is conducted is a decision of the
+/// server and is outside of the scope of this protocol. Likewise, the order in
+/// which the supplied certificate chain is validated and the specification of
+/// trust anchors used to terminate validation are also controlled by the server.
+pub(crate) async fn validate_operation(
+    kms: &KMS,
+    request: Validate,
+    user: &str,
+    params: Option<&ExtraDatabaseParams>,
+) -> KResult<ValidateResponse> {
+    trace!("Validate: {:?}", request);
+    let certificates = match (request.unique_identifier, request.certificate) {
+        (None, None) =>
+        // None uid, None cert case
+        {
+            return KResult::Ok(ValidateResponse {
+                validity_indicator: ValidityIndicator::Unknown,
+            })
+        }
+        (None, Some(certificates)) => {
+            // None uid, Some cert case
+            KResult::Ok(certificates)
+        }
+        (Some(unique_identifiers), None) =>
+        // None uid, Some cert case
+        {
+            certificates_by_uid(unique_identifiers, kms, user, params).await
+        }
+        (Some(unique_identifiers), Some(certificates)) =>
+        // None uid, Some cert case
+        {
+            let mut certificates_uid =
+                certificates_by_uid(unique_identifiers, kms, user, params).await?;
+            let mut certificates = certificates.clone();
+            certificates.append(&mut certificates_uid);
+            KResult::Ok(certificates.clone())
+        }
+    }?;
+
+    // Indexing Certificate Chain
+    let hm_certificates =
+        &mut index_certificates(&certificates, &mut HashMap::<Vec<u8>, u8>::new())?;
+    println!("len certificates : {}", certificates.len());
+    println!("len hm_certificates : {}", hm_certificates.len());
+
+    // Getting root certificate from indexing
+    let root_idx = if let Some(root_idx) = hm_certificates.get(&HEAD.to_vec()) {
+        *root_idx
+    } else {
+        return KResult::Err(KmsError::from(KmipError::ObjectNotFound(
+            "The certificate chain has no root".to_string(),
+        )))
+    };
+    let root_cert = certificates
+        .get(root_idx as usize)
+        .expect("root must exist");
```

Comment:
please use `ok_or_else` combined to an Error to avoid an server panic/shutdown (and remove all `expect`)